python3/bin/perfmon

```diff
@@ -0,0 +1,1620 @@
+            else:
+                # check if the config has changed, e.g. by XenCenter
+                host_mon.refresh_config()
+
+            # List of SRs present in rrd_updates
+            sr_uuid_list = rrd_updates.get_uuid_list_by_objtype("sr")
+            print_debug("sr_uuid_list = %s" % sr_uuid_list)
+
+            # Remove monitors for SRs no longer listed in the rrd_updates page
+            # We use .pop() inside the loop, use list(dict_var.keys()):
+            for uuid in list(sr_mon_lookup.keys()):
+                if uuid not in sr_uuid_list:
+                    sr_mon_lookup.pop(uuid)
+            # Create monitors for SRs that have just appeared in rrd_updates page
+            for uuid in sr_uuid_list:
+                if uuid not in sr_mon_lookup:
+                    sr_mon_lookup[uuid] = SRMonitor(uuid)
+                else:
+                    sr_mon_lookup[uuid].refresh_config()
+
+            # Go through each vm_mon and update it using the rrd_udpates
+            # this may generate alarms
+            for vm_mon in vm_mon_lookup.values():
+                vm_mon.process_rrd_updates(rrd_updates, session)
+
+            # Ditto for the host_mon
+            if host_mon:
+                host_mon.process_rrd_updates(rrd_updates, session)
+
+            # And for the sr_mons
+            for sr_mon in sr_mon_lookup.values():
+                sr_mon.process_rrd_updates(rrd_updates, session)
+
+        except ConnectionRefusedError as e:
+            # "Connection refused[111]"
+            # this happens when we try to restart session and *that* fails
+            time.sleep(2)
+            log_err(
+                "caught connection refused error: (%s) - restarting XAPI session"
+                % str(e)
+            )
+            restart_session = True
+        except urllib.error.HTTPError as e:
+            if e.code in (401, 500):
+                # Error getting rrd_updates: 401=Unauthorised, 500=Internal
+                # start new session
+                log_err("caught http.error: (%s) - restarting XAPI session" % str(e))
+                restart_session = True
+            else:
+                # Don't know why we got this error - crash, die and look at logs later
+                raise
+        except OSError as e:
+            # This happens if we send messages or
+            # read other-config:perfmon after xapi is restarted
+            log_err("caught connection error: (%s) - restarting XAPI session" % str(e))
+            restart_session = True
+
+        runs += 1
+        if maxruns is not None and runs >= maxruns:
+            break
+
+        # Force collection of cyclically referenced objects cos we don't
+        # trust GC to do it on its own
+        gc.collect()
+
+        # Sleep for interval + dither, exiting early if we recv a cmd
+        timeout = rand(interval, interval + dither)
+        cmdsock.settimeout(timeout)
+        try:
+            cmd = cmdsock.recv(cmdmaxlen)
```

Comment:
Would this be a bytes object?

Reply:
It is indeed a `bytes` object.
Nice! Another `bytes-str` issue. 
This code should've been covered, but because of the [issue ](https://github.com/xapi-project/xen-api/pull/5528)you fixed days ago, the XenRT test failed.
Will modify it.